### ISSUE TITLE
Importing 3rd Party SCSS

### ISSUE THREAD

**Issue by gonzofish** [gonzofish/angular-library-set#37]
Follow-on from [this comment](https://github.com/gonzofish/angular-library-set/issues/27#issuecomment-304660371) on #27

**Comment by tommueller** [comment]
@gonzofish just played around with this and I cannot use any `@import` within the scss files. I get the error also when I add the file that should be imported into the components folder directly.

Would really like to help you out with a PR, but I do not really understand this project in the first place, sorry :man_facepalming:

**Comment by gonzofish** [comment]
After some digging around and testing, I _might_ have a fix (based on [this comment](https://github.com/webpack-contrib/extract-text-webpack-plugin/issues/234#issuecomment-289265024)), although I can't really tell.

In my sample project I `@import`ed two local files (1 css, 1 scss) and one of the Angular Material prebuilt themes CSS. I also created a node module and imported that into my component and it worked. So like I said, I _think_ it will work.

I'll close this issue when the change has been published.

**Comment by gonzofish** [comment]
The original issue was finally found...we're using [node-sass](https://github.com/sass/node-sass) to convert SASS files to CSS before inlining them in components. Part of the problem is that node-sass doesn't really do `@import` well. You have to specify an `importer` attribute when rendering the SASS to CSS, like:

```javascript
            importer: (url, prev, done) => {
                if (/^~/.test(url)) {
                    url = path.resolve('node_modules', url.slice(1));
                } else {
                    url = path.resolve(srcDir, url);
                }

                return { contents: fs.readFileSync(url, 'utf8') };
            }
```

Basically we're going to look for a leading `~` and, if so, we'll replace that leading `~` with `node_modules` and read the file from there. Otherwise, we make the URL relative to the component's directory (this should let `../` work too) and read the file from there.

**Comment by gonzofish** [comment]
This is all going to be happening in the [`tasks/inline-resources.js` file](https://github.com/gonzofish/angular-library-set/blob/master/commands/initial/templates/tasks/inline-resources.js#L76) at line 76.

**Comment by gonzofish** [comment]
Tested locally and seems to work.

[Here's the release](https://github.com/gonzofish/angular-library-set/releases/tag/1.0.0-alpha.15)

**Comment by tommueller** [comment]
Thank you so much for your effort! It kind of works for me, but not really ...

I _can_ successfully import .css files now, but _cannot_ import .scss-files. I still get the same error with .scss-files. I forked the repo, because I thought it's probably an easy fix, but I saw that you did include .scss everywhere. Did you test it with .scss-files?

Cheers,
Tom

**Comment by gonzofish** [comment]
Hi Tom -

I did test it with both CSS & SCSS files. I'm obviously missing something here...your import looks like:

```scss
@import '~package/dist/scss/imports';
```

Correct? Imports, I take it, is actually `imports.scss`?

**Comment by tommueller** [comment]
Here is a list with options I tested and with the result (with an css and scss in the parent folder, and in the node_modules/package folder:

-  @import '../imports'; (to import imports.scss) -> _not_ working 
-  @import '../imports.scss'; -> _not_ working
-  @import '../imports.1'; (to import _imports.1.scss by partials naming-convention) -> _not_ working
-  @import '../imports.1.scss'; -> _not_ working
-  @import '~package/dist/scss/imports.scss'; -> _not_ working
-  @import '~/package/dist/scss/imports.scss'; -> _not_ working
-  @import '~package/dist/scss/imports'; -> _not_ working
-  @import '~/package/dist/scss/imports'; -> _not_ working
-  @import '../importCss.css'; -> _working_
-  @import '../importCss'; -> _not_ working
-  @import '~package/dist/css/ifaceplate.min.css'; -> _working_

so basically everything with scss is not working for me and with css it does ...

**Comment by gonzofish** [comment]
Thanks for doing this work. I haven't had trouble with files with an `.scss` extension, but did have issues with files with no extension (which I'm working on fixing).

Can you look in `tasks/inline-resources.js` line 76 and see if it looks like [this one in the repo](https://github.com/gonzofish/angular-library-set/blob/master/commands/initial/templates/tasks/inline-resources.js#L76)?

**Comment by gonzofish** [comment]
Also, I'd never used partials in Sass before, so that'll take some work as well

**Comment by tommueller** [comment]
> Can you look in tasks/inline-resources.js line 76 and see if it looks like this one in the repo?

Yes, looks exactly like it!

**Comment by gonzofish** [comment]
Ok...thanks. Looking into it.

**Comment by gonzofish** [comment]
Do the partials work when you use `ngl serve`?

**Comment by tommueller** [comment]
> Do the partials work when you use ngl serve?

yes

**Comment by gonzofish** [comment]
@tommueller just a heads up, on a local project, it seems to be working for both serve & build. Here's the sample Sass file I have:

```sass
@import "~fake-node-lib/fake.css";
@import "~fake-node-lib/fake-scss";
@import "local.css";
@import "../partial";

.test {
    color: $text-color;
    height: 200px;
    width: 200px;
}
```

The variable `$text-color` gets imported from `_partial.scss` a level above and other stylings get pulled in from the other stylesheets.

**Comment by gonzofish** [comment]
I've published 1.0.0-alpha.16 which tries to solve this problem. I'll leave it open until we get a confirmation from you @tommueller

**Comment by tommueller** [comment]
Thanks a lot @gonzofish 

I will only be able to check on this on Monday. Will get back to you as soon as I tested it!

**Comment by gonzofish** [comment]
Not a problem. Thank you for being patient and working with me on this.

**Comment by tommueller** [comment]
@gonzofish great this works for me now! Thank you very much!

I did have to initialize a whole new project though, could not get the original one to work after upgrading to alpha.16 ...

I also now get this message in the debug-output of VSCode when building the project:
```
[Error - 07:35:49] ImportResolveExtension - There was an error during reprocessing changed files.
	Data:	{ Error: ENOENT: no such file or directory, open '/home/tom/workspace_angular/interfaceMA-uiElements/build/index.ts'
    at Object.fs.openSync (fs.js:558:18)
    at Object.module.(anonymous function) [as openSync] (ELECTRON_ASAR.js:173:20)
    at Object.fs.readFileSync (fs.js:468:33)
    at Object.fs.readFileSync (ELECTRON_ASAR.js:506:29)
    at filePathes.map.o (/home/tom/.vscode/extensions/rbbit.typescript-hero-0.13.2/out/src/common/ts-parsing/TypescriptParser.js:24:65)
    at Array.map (native)
    at TypescriptParser.<anonymous> (/home/tom/.vscode/extensions/rbbit.typescript-hero-0.13.2/out/src/common/ts-parsing/TypescriptParser.js:24:18)
    at Generator.next (<anonymous>)
    at /home/tom/.vscode/extensions/rbbit.typescript-hero-0.13.2/node_modules/tslib/tslib.js:96:75
    at Object.__awaiter (/home/tom/.vscode/extensions/rbbit.typescript-hero-0.13.2/node_modules/tslib/tslib.js:92:16)
  errno: -2,
  code: 'ENOENT',
  syscall: 'open',
  path: '/home/tom/workspace_angular/interfaceMA-uiElements/build/index.ts' }
```

**Comment by gonzofish** [comment]
Great to hear it worked. Do you remember what exactly wasn't working?

**Comment by gonzofish** [comment]
Did you remove the stack trace? I see it in the email but not the issue here...

**Comment by tommueller** [comment]
In my old project I still got the same error as posted here: https://github.com/gonzofish/angular-library-set/issues/27#issuecomment-304660371

I removed the stack trace, that error was unrelated (caused by a VSCode-plugin).

**Comment by gonzofish** [comment]
I know this is a silly question, but you did upgrade the older project? The reason for that error is that it is saying it cannot find the file specified. On a re-init'd project (sorry, that's the "update" method for now), this should only happen with the `@import`-ing of a `.css` file that is missing.

**Comment by tommueller** [comment]
Yes, I did upgrade ;) or at least I did this:

- update the version of angular-librarian to alpha-16 in package.json
- run `yarn upgrade`

I think that should've done it, right?

**Comment by gonzofish** [comment]
That would install alpha.16, but you'd still need to run angular-librarian's `init` to re-scaffold the project using the latest. I guess since I know at least 1 person is actively using this I should investigate a better migration/update strategy from version-to-version.

**Comment by tommueller** [comment]
Okay, I did not know and do that!

Wasn't a problem, because it was a one component library so far, but very good to know for the future!

**Comment by gonzofish** [comment]
Ok so this is officially closed, but has spawned a new issue for better migration: #38
